### PR TITLE
[PM-15061] extract encryptors from generator service

### PR DESCRIPTION
**Pull request by audreyality**
## 🎟️ Tracking

https://bitwarden.atlassian.net/browse/PM-15061

## 📔 Objective

Extract encryptors from generator service so they can be reused across the tools codebase.

## 🦖 System Evolution

A new set of encryptors that distinguishes key slots from key instances is in the works. These encryptors also have specialized key rotation interfaces to enable integration with key services. 

This PR introduces a parallel set of interfaces (`LegacyEncryptorProvider` et. al.) as a step towards that new system. Like the legacy generator services before it, the legacy encryptor interfaces arrive deprecated so that the impact of importing them is clear--porting to the new interfaces will be required once they become available.

## 🦮 Reviewer guidelines

- 👍 (`:+1:`) or similar for great changes
- 📝 (`:memo:`) or ℹ️ (`:information_source:`) for notes or general info
- ❓ (`:question:`) for questions
- 🤔 (`:thinking:`) or 💭 (`:thought_balloon:`) for more open inquiry that's not quite a confirmed issue and could potentially benefit from discussion
- 🎨 (`:art:`) for suggestions / improvements
- ❌ (`:x:`) or ⚠️ (`:warning:`) for more significant problems or concerns needing attention
- 🌱 (`:seedling:`) or ♻️ (`:recycle:`) for future improvements or indications of technical debt
- ⛏ (`:pick:`) for minor or nitpick changes

### CODE REVIEW

File: libs/common/src/tools/cryptography/organization-encryptor.abstraction.ts [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+import { EncString } from "../../platform/models/domain/enc-string";
+
+/** An encryption strategy that protects a type's secrets with
+ *  organization-specific keys. This strategy is bound to a specific organization.
+ */
+export abstract class OrganizationEncryptor {
+  /** Identifies the user bound to the encryptor. */
+  readonly organizationId: OrganizationId;
+
+  /** Protects secrets in `value` with a user-specific key.
```
Comment (quexten) [inline libs/common/src/tools/cryptography/organization-encryptor.abstraction.ts:14]: "User specific key" should this be org specific key? Orgs currently just have 1 org wide key.

File: libs/common/src/tools/cryptography/organization-key-encryptor.spec.ts [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+    it("should throw if secret was not supplied", async () => {
+      const encryptor = new OrganizationKeyEncryptor(anyOrgId, encryptService, orgKey, dataPacker);
+
+      await expect(encryptor.decrypt(null)).rejects.toThrow("secret cannot be null or undefined");
+      await expect(encryptor.decrypt(undefined)).rejects.toThrow(
+        "secret cannot be null or undefined",
+      );
+    });
+
+    it("should declassify a decrypted packed value using the user's key", async () => {
```
Comment (quexten) [inline libs/common/src/tools/cryptography/organization-key-encryptor.spec.ts:113]: Should this be orgs key?

Comment (audreyality) [reply libs/common/src/tools/cryptography/organization-key-encryptor.spec.ts]: Addressed in 2b806da07.

File: libs/common/src/tools/cryptography/organization-encryptor.abstraction.ts (hunk 1 above)
Comment (audreyality) [reply libs/common/src/tools/cryptography/organization-encryptor.abstraction.ts]: Addressed in 2b806da07

File: libs/common/src/tools/rx.ts [hunk 3]
```diff
@@ -15,8 +15,58 @@ import {
...
   concatMap,
+  startWith,
+  pairwise,
 } from "rxjs";
 
+function identity(value: any): any {
+  return value;
+}
+
+function expectedAndActualValue(expectedValue: any, actualValue: any) {
```
Comment (djsmith85) [inline libs/common/src/tools/rx.ts:26]: ⚠️ Please add some documentation

Comment (audreyality) [reply libs/common/src/tools/rx.ts:28]: Addressed in 65fbe1873.

File: libs/common/src/tools/rx.ts [hunk 4]
```diff
@@ -15,8 +15,58 @@ import {
   takeUntil,
   withLatestFrom,
   concatMap,
+  startWith,
+  pairwise,
 } from "rxjs";
 
+function identity(value: any): any {
```
[the remark this answers is not in the document]
Comment (audreyality) [reply libs/common/src/tools/rx.ts:23]: Addressed in 65fbe1873.

**Review by djsmith85 — APPROVED** [review]

**Review by quexten — APPROVED** [review]
Looks good from my side.

**Review by quexten — APPROVED** [review]

**Review by djsmith85 — APPROVED** [review]

**Pull request merged**